producer/writer/retryable_connection.go

```diff
@@ -0,0 +1,259 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+const (
+	defaultRetryForever = true
+)
+
+var (
+	errConnNotInitialized = errors.New("connection not initialized")
+)
+
+type retryableConnectionMetrics struct {
+	resetConn         tally.Counter
+	resetError        tally.Counter
+	connectError      tally.Counter
+	setKeepAliveError tally.Counter
+}
+
+func newRetryableConnectionMetrics(scope tally.Scope) retryableConnectionMetrics {
+	return retryableConnectionMetrics{
+		resetConn:         scope.Counter("reset-conn"),
+		resetError:        scope.Counter("reset-conn-error"),
+		connectError:      scope.Counter("connect-error"),
+		setKeepAliveError: scope.Counter("set-keep-alive-error"),
+	}
+}
+
+type connectFn func(addr string) (net.Conn, error)
+
+type retryableConnection struct {
+	addr            string
+	retrier         retry.Retrier
+	opts            Options
+	resetDelayNanos int64
+	logger          log.Logger
+
+	connLock       sync.RWMutex
+	conn           net.Conn
+	w              *bufio.Writer
+	r              *bufio.Reader
+	lastResetNanos int64
+	initialized    bool
+	closed         *atomic.Bool
+	doneCh         chan struct{}
+	resetCh        chan struct{}
+	wg             sync.WaitGroup
+	m              retryableConnectionMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+// newRetryableConnection creates a connection that is automatically
+// reconnected on errors.
+func newRetryableConnection(
+	addr string,
+	opts Options,
+) *retryableConnection {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	c := &retryableConnection{
+		addr:            addr,
+		retrier:         retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		opts:            opts,
+		resetDelayNanos: int64(opts.ConnectionResetDelay()),
+		logger:          opts.InstrumentOptions().Logger(),
+		w:               bufio.NewWriterSize(nil, opts.ConnectionWriteBufferSize()),
+		r:               bufio.NewReaderSize(nil, opts.ConnectionReadBufferSize()),
+		lastResetNanos:  0,
+		initialized:     false,
+		closed:          atomic.NewBool(false),
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newRetryableConnectionMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	c.connectFn = c.connectOnce
+	if err := c.tryResetWithConnectFn(c.connectFn); err != nil {
```

Comment:
nit: call this `resetWithConnectFn`. `tryXXX` conventionally returns a boolean rather than an error.